### PR TITLE
Add optional anonymous Tempo usage reporting

### PR DESCRIPTION
**Pull request by zalegrala**
**What this PR does**:

Here we implement an approach from the Loki squad for sending anonymous usage information to Grafana Labs to better understand the uses in the wild.

Fixes https://github.com/grafana/tempo-squad/issues/81
Related https://github.com/grafana/loki/issues/5062

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/usagestats/config.go
Comment (joe-elliott) [inline pkg/usagestats/config.go:1]: have we pursued finding a shared location for this code instead of forking it?

@mattmendick

Comment (zalegrala) [reply pkg/usagestats/config.go:1]: There was a conversation in a tech leads channel about creating a place for some of this, since folks are noticing the duplication.  There is room to split this up some and move pieces to a shared location, but some items seem somewhat specific to tempo.  The items that I'm thinking of are the storage clients to the various backends and their associated configurations as well as some error handling.

A `backend-libs` repo could exist and start moving components.  Perhaps some coordination makes sense.

Comment (mattmendick) [reply pkg/usagestats/config.go:1]: I think forking it is a good short term solution this minute, but I would really like to see a shared library for this kind of thing. If there is time to create this and start contributing things to it, I think that would be a win.

Comment (mattmendick) [reply pkg/usagestats/config.go:1]: I'll leave it to y'all to discuss if there is time for that added overhead, but it would be a net positive on the databases group to take the time to do this.

Comment (zalegrala) [reply pkg/usagestats/config.go:1]: There doesn't seem to be much appetite from the leads for sharing this code among the backend databases, at least for now.  We could revisit at a later time, but not sure when we would.

File: modules/distributor/receiver/shim.go
Comment (zalegrala) [inline modules/distributor/receiver/shim.go:109]: I'm unsure if the defaults here will be 0, or if we should explicitly set them to 0.

**Review by joe-elliott — COMMENTED** [review]
some thoughts

**Comment by zalegrala** [comment]
That's great, thanks for the feedback.

File: modules/distributor/receiver/shim.go
Comment (mdisibio) [reply modules/distributor/receiver/shim.go:117]: > is there a way to associate a string label with this "stat"? so instead of a bunch of individual stats we could says receiverStats.Type(<name>).Set(1) or something?

Discussed offline a bit, and the idea here is to centralize the stats more concretely in the usagestats package. Instead of calling something like `usagestats.NewInt("feature_enabled_search")`, `int.Set(1)` throughout the code base, this could be `usagestats.SetFeatureEnabledSearch(1)`.  The `SetFeatureEnabledSearch` method would call NewInt/Set.

Thoughts?

File: cmd/tempo/app/app.go
Comment (zalegrala) [reply cmd/tempo/app/app.go:61]: So you like `statFeatureEnabledMetricsGenerator`?

File: cmd/tempo/app/modules.go
Comment (zalegrala) [reply cmd/tempo/app/modules.go:334]: A couple things come to mind.  Having a client instance dedicated to this module seems like a good way to avoid contention/blocking on client use in other parts of the system.  I don't know this to be the case with our clients, but just something I've seen.  Also, since all of the main components will be running this code that relies on the backend, it didn't seem right to require all components to require the ingester just to get the client.

File: go.mod
Comment (zalegrala) [reply go.mod:97]: Not sure, but I'll `go mod tidy` a little harder and see what happens.

File: modules/distributor/receiver/shim.go
Comment (zalegrala) [reply modules/distributor/receiver/shim.go:117]: The current design of the `usagesats` package is such that it doesn't care at all about what stats are being set throughout the code base.  To create some helper methods in the package _might_ help readability on the implementation side, but I'm thinking the separation of concerns here is somewhat nice.  If the variables names with a `Set()` method are ugly to look at, we could also make package local fucntions like `setFeatureEnabledSearch()` that would make the necessary calls in. 

With a package variable like `featureEnabledSearch`, calling `featureEnabledSearch.Set(1)` feels almost as readable as `usagestats.SetFeatureEnabledSearch()`, the difference being that now the usagestats needs modification for each stat we want to include.  Is there another advantage of moving the stats to the usagestats package here I'm not thinking of?

File: pkg/usagestats/reporter.go
Comment (zalegrala) [reply pkg/usagestats/reporter.go:192]: Perhaps we could avoid deleting the object, but checking for the presence of an object seems a little simpler than trying to read a busted file.  We could make a new interface though if we have concerns about reuse.

File: tempodb/backend/raw.go
Comment (zalegrala) [reply tempodb/backend/raw.go:33]: We could move this to a new interface to address the concern.  All of the backends would keep the current implementation, and just conform to an additional interface if that's what we want.  Extending the interface here seemed the cleanest to me considering that we would need to extend the implementations for each backend to meet the new functionality.

File: pkg/usagestats/reporter.go
Comment (zalegrala) [reply pkg/usagestats/reporter.go:1]: I'll include a header in this file with the sha that this code was forked from.  The commit message has attribution also.

File: pkg/usagestats/reporter.go
Comment (annanay25) [inline pkg/usagestats/reporter.go:266]: Should we turn this into info level?

**Review by annanay25 — COMMENTED** [review]
Some concerns as discussed offline:
1. I'm a little skeptical to adding a dependency on the backend storage to the usage reporting module - not only because of the complexity of running the module but also an unexpected failure scenario with the backend that might result in a component crash. I believe we could hold the token in the memberlistkv for components, in which case it will only ever be recreated if **all** components are wiped and restarted - at which point we might as well call it a new cluster.
2. With the additional dependency on the backend, components like the distributor which initially had no dependency on the backend will have to be configured with the read token from GCS/S3/Azure. This is a breaking change and needs to be communicated clearly and updated in all docs.
3. Schema changes - I'm not sure what happens if we change the usage report schema, does an old payload get rejected with a 400? Or do we accept with 200 ok but drop it internally? How hard is it to change the associated dashboards?
4. The backoff to send data / wait for the token to be created should definitely be made configurable. In a large cluster with a 100 components, it can result in pretty spikey memberlist traffic if all components query the kv store every second

File: cmd/tempo/app/app.go
Comment (joe-elliott) [reply cmd/tempo/app/app.go]: like the changes i'm seeing here

File: cmd/tempo/app/modules.go
Comment (joe-elliott) [reply cmd/tempo/app/modules.go:334]: So, tempodb should have no issues with concurrency. However, I do see now that the standard tempodb interfaces do not support what you want and we would not want to adjust them for this feature.  I also see now that we'd have to start up a storage module for the distributor, which we definitely would not want.

Agree this is likely the best way to do things.

I do think a small improvement would be to move this code to the `/tempodb/backend` package which we call like:

`backend.New(t.cfg.StorageConfig.Trace) (reader, writer, compactor, err)`

it could then be reused by `tempodb.New()`

File: docs/tempo/website/configuration/_index.md
Comment (joe-elliott) [reply docs/tempo/website/configuration/_index.md:1114]: What might make even more sense it to provide an endpoint (`/status/usage-stats`) that just dumps all of the information included in usage-stats. Is this even possible?

I don't consider that a requirement for this PR. Can you open an issue to add the new status endpoint and then we can document it here once it exists?

File: pkg/usagestats/reporter.go
Comment (joe-elliott) [reply pkg/usagestats/reporter.go:194]: Extending an interface is generally costly and should be questioned. I think we're getting at: What is the purpose of this line and how does it interact with the process of generating/sharing a seed. Is there a way to still do that reliably w/o adding the delete function?

File: tempodb/backend/raw.go
Comment (joe-elliott) [reply tempodb/backend/raw.go:33]: This conversation is a duplicate of https://github.com/grafana/tempo/pull/1481/files#r911031066

Going to mark this resolved and continue the question there.

File: modules/distributor/receiver/shim.go
Comment (joe-elliott) [reply modules/distributor/receiver/shim.go]: I'm fine with either way. I like @mdisibio's suggestion to consolidate the stats b/c it will help users of the software quickly see what we are reporting. I don't consider this a blocker.

File: docs/tempo/website/configuration/_index.md
Comment (knylander-grafana) [inline docs/tempo/website/configuration/_index.md:1124]:
```suggestion
The reporting is controlled by a configuration option. You can disable the
automatic reporting of this generic information using the following configuration:
```

Comment (knylander-grafana) [inline docs/tempo/website/configuration/_index.md:1124]:
```suggestion
By default, Tempo will report anonymous usage data about the shape of a deployment 
to Grafana Labs. This data is used to determine how common
the deployment of certain features are, or if a feature flag has been enabled,
replication factor or compression levels, etc.
 
Reporting is controlled by a configuration option. You can disable the
automatic reporting of this generic information using the following configuration:
```

**Review by knylander-grafana — APPROVED** [review]
Made a minor suggestion to the text.

File: docs/tempo/website/configuration/_index.md
Comment (zalegrala) [reply docs/tempo/website/configuration/_index.md:1114]: Nice, I like that idea.  I'll file an issue.

Comment (zalegrala) [reply docs/tempo/website/configuration/_index.md:1114]: https://github.com/grafana/tempo/issues/1569

File: modules/distributor/receiver/shim.go
Comment (zalegrala) [reply modules/distributor/receiver/shim.go]: Fair enough.  I see that it also might prevent sharing the package in the future, but I don't know how much to hold on to that idea.

File: cmd/tempo/app/modules.go
Comment (zalegrala) [reply cmd/tempo/app/modules.go:334]: I like this idea of centralizing the New for the backend client creation.  At first pass, this creates a dependency cycle, but perhaps after I address some of the other comments here we might be able to get out of the loop.

File: pkg/usagestats/reporter.go
Comment (zalegrala) [reply pkg/usagestats/reporter.go]: I've made an update to remove the interface extension and created a new error to indicate that we can replace the seed file during start up.  I'll run it locally shortly to confirm, the the test has also been updated.

Thoughts?

File: modules/distributor/receiver/shim.go
Comment (zalegrala) [reply modules/distributor/receiver/shim.go]: I think we could move all the variables to `modules.go` if we wanted one place to look.  We mostly need access to the config, which could be done in a helper on the Server.  Hows that sound?  I think that might smooth out a dependency loop for the backend creation also when trying to implement `backend.New()` for shared use.

File: cmd/tempo/app/modules.go
Comment (zalegrala) [reply cmd/tempo/app/modules.go:334]: Yeah, let's refactor the backend later.

File: pkg/usagestats/reporter.go
Comment (zalegrala) [reply pkg/usagestats/reporter.go]: I've updated the implementation to avoid the delete call and the extension to the interface.

**Review by annanay25 — APPROVED** [review]
Left a final few comments but approving to unblock once done!

File: pkg/usagestats/config.go
Comment (zalegrala) [reply pkg/usagestats/config.go]: This has been updated.  Hows that?

File: pkg/usagestats/reporter_test.go
Comment (zalegrala) [reply pkg/usagestats/reporter_test.go:214]: I don't think so, because only the ingesters will require the k/v store.  This will get checked when the ingester is started during the call to `running()`.

**Pull request merged**
